src/Compiler/Checking/CheckExpressions.fs

```diff
@@ -7332,6 +7332,33 @@ and TcInterpolatedStringExpr cenv (overallTy: OverallTy) env m tpenv (parts: Syn
             // Type check the expressions filling the holes
             let fillExprs, tpenv = TcExprsNoFlexes cenv env m tpenv argTys synFillExprs
 
+            // If all fill expressions are strings and there is less then 5 parts of the interpolated string total
+            // then we can use System.String.Concat instead of a sprintf call
+            if isString && (parts.Length < 5) && (argTys |> List.forall (isStringTy g)) then
+                let rec f xs ys acc =
+                    match xs with
+                    | SynInterpolatedStringPart.String(s, m)::xs ->
+                        let sb = System.Text.StringBuilder(s).Replace("%%", "%")
+                        f xs ys ((mkString g m (sb.ToString()))::acc)
+                    | SynInterpolatedStringPart.FillExpr(_, _)::xs ->
+                        match ys with
+                        | y::ys -> f xs ys (y::acc)
+                        | _ -> error(Error((0, "FOOBAR"), m)) // TODO XXX wrong error
+                    | _ -> acc
+
+                let args = f parts fillExprs [] |> List.rev
+                assert (args.Length = parts.Length)
+                if args.Length = 4 then
```

Comment:
Was number 4 chosen because that is the max. number where `String.concat` has a dedicated method overload?

Reply:
Yeah, see this https://sharplab.io/#v2:DYLgZgzgNALiCWwoBMQGoA+BlAnhGApgLYB0WMATvAHYDmJAwgPbUDGAhjABQBE7PUAAQ8ARgOGtxPZDwCUQA=== vs https://sharplab.io/#v2:DYLgZgzgNALiCWwoBMQGoA+BlAnhGApgLYB0WMATvAHYDmJAwgPbUDGAhjABQBE7PUAAQ8ARgOGtxPZFII8AlEA=